kernel_gateway/services/kernels/pool.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from tornado.locks import Semaphore
+from tornado import gen
+
+
+class KernelPool():
+    '''
+    A class to maintain a pool of kernel and control access to the individual kernels.
+    Kernels are protected by a borrower/lender pattern.
+    '''
+    def __init__(self, prespawn_count, kernel_manager):
+        self.kernel_clients = {}
+        self.on_recv_funcs = {}
+        self.kernel_manager = kernel_manager
+        self.pool_index = 0
+        self.kernel_pool = []
+        self.kernel_semaphore = Semaphore(prespawn_count)
+
+        for _ in range(prespawn_count):
+            kernel_id = kernel_manager.start_kernel()
+            self.kernel_clients[kernel_id] = kernel_manager.get_kernel(kernel_id).client()
+            self.kernel_pool.append(kernel_id)
+            iopub = self.kernel_manager.connect_iopub(kernel_id)
+            iopub.on_recv(self.create_on_reply(kernel_id))
+
+    @gen.coroutine
+    def acquire(self):
```

Comment:
Queueing up the requests "worked", but we decided to still go this route because it helps serves requests faster. Best explained with an example. Imagine two kernels and three request:
1. Request 1 executes on kernel 1 and takes 30 seconds
2. Request 2 executes on kernel 2 and takes 1 second
2. Request 3 executes on kernel 1 and takes 1 second

If we simply incremented the counter, request 3 will need to wait for request 1 to finish. This approach will server request 3 after request 2 is finished.